### PR TITLE
Fixes: #1298 - Add SnipeIT source

### PR DESCRIPTION
**Pull request by chandanchowdhury**
### Summary

Adding new intel module for SnipeIT source.

### Related issues
#1298
___
Read through our [developer docs](https://lyft.github.io/cartography/dev/developer-guide.html)

- [x] PR Title starts with "Fixes: [issue number]"

If you are modifying or implementing a new intel module
- [x] Update the [schema](https://github.com/lyft/cartography/tree/master/docs/root/modules) and [readme](https://github.com/lyft/cartography/blob/master/docs/schema/README.md)
- [x] Use our NodeSchema [data model](https://lyft.github.io/cartography/dev/writing-intel-modules.html#defining-a-node)
- [x] Use specialized functions `get_`, `transform_`, `load_`, and `cleanup_` functions
- [x] Add [tests](https://lyft.github.io/cartography/dev/writing-intel-modules.html#making-tests)
  - [x] Unit tests: Test your `transform_` function with sample data (Not applicable)
  - Integration tests
    - [x] Use our test [helper functions](https://github.com/lyft/cartography/blob/master/tests/integration/util.py)
    - [x] Test a cleanup job

### CODE REVIEW

**Comment by chandanchowdhury** [comment]
Review request

@achantavy

File: cartography/intel/snipeit/asset.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    return response["rows"]
+
+
+@timeit
+def load_assets(
+    neo4j_session: neo4j.Session,
+    common_job_parameters: Dict,
+    data: List[Dict[str, Any]],
+) -> None:
+    logger.debug(data[0])
```
Comment (achantavy) [inline cartography/intel/snipeit/asset.py:31]: Remove debug print to the data

File: cartography/intel/snipeit/util.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+# Connect and read timeouts of 60 seconds each; see https://requests.readthedocs.io/en/master/user/advanced/#timeouts
+_TIMEOUT = (60, 60)
+
+
+@timeit
+def call_snipeit_api(api_and_parameters: str, base_uri: str, token: str) -> Dict[str, Any]:
+    uri = base_uri + api_and_parameters
+    try:
+        logger.debug(
+            "SnipeIT: Get %s", uri,
```
Comment (achantavy) [inline cartography/intel/snipeit/util.py:19]: Double checking, this won't have a cred in it right?

File: cartography/intel/snipeit/util.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+from cartography.util import timeit
+
+logger = logging.getLogger(__name__)
+# Connect and read timeouts of 60 seconds each; see https://requests.readthedocs.io/en/master/user/advanced/#timeouts
+_TIMEOUT = (60, 60)
+
+
+@timeit
+def call_snipeit_api(api_and_parameters: str, base_uri: str, token: str) -> Dict[str, Any]:
```
Comment (achantavy) [inline cartography/intel/snipeit/util.py:15]: Non-block to consider for the future: use something like backoff to do autoretry


Example
```
import backoff

@backoff.on_exception(
    wait_gen=backoff.expo,
    exception=RuntimeError,
    max_tries=5,
    on_backoff=backoff_handler,
)
```

File: docs/root/modules/snipeit/config.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+## SnipeIT Configuration
+
+.. _SnipeIT_config:
+
+Follow these steps to analyze SnipeIT users and assets in Cartography.
+
+1. Prepare an API token for SnipeIT
+    1. Follow [SnipeIT documentation](https://snipe-it.readme.io/reference/generating-api-tokens) to generate a API token.
+    1. Populate `SNIPEIT_TOKEN` environment variable with the API token. Alternately, you can pass a different environment variable name containing the API token
```
Comment (achantavy) [inline docs/root/modules/snipeit/config.md:9]: Not familiar with SnipeIT, but is there a specific permission that the token needs? If so, should include that in documentation for the user.

File: docs/root/modules/snipeit/schema.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+|manufacturer | Manufacturer of the asset |
+|model | Model of the device|
+|serial | Serial number of the asset|
+
+#### Relationships
+
+- All SnipeIT users and asset are linked to a SnipeIT Tenant
+
+    ```cypher
+    (:SnipeitUser)-[:USER]->(:SnipeitTenant)
```
Comment (achantavy) [inline docs/root/modules/snipeit/schema.md:44]: I think the direction is actually reversed here

File: docs/root/modules/snipeit/schema.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    ```
+
+    ```cypher
+    (:SnipeitAsset)-[:ASSET]->(:SnipeitTenant)
```
Comment (achantavy) [inline docs/root/modules/snipeit/schema.md:48]: I think the direction is actually reversed here too

File: tests/integration/cartography/intel/snipeit/test_snipeit_assets.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        'snipeitTenant',
+        ['id'],
+    )
+
+    # Make sure the expected assets are created
+    expected_nodes = {
+        (1373, "C02ZJ48XXXXX"),
+        (1372, "72ec94a8-b6dc-37f1-b2a9-0907806e8db7"),
+    }
+    assert check_nodes(
```
Comment (achantavy) [inline tests/integration/cartography/intel/snipeit/test_snipeit_assets.py:56]: Nit: could rewrite as 

```python
assert check_nodes(
        neo4j_session,
        "SnipeitAsset",
        ["id", "serial"],
    ) == {
        (1373, "C02ZJ48XXXXX"),
        (1372, "72ec94a8-b6dc-37f1-b2a9-0907806e8db7"),
}
```

to save a couple lines

**Review by achantavy — CHANGES_REQUESTED** [review]
Minor nits but looks good

File: docs/root/modules/snipeit/config.md (hunk 4 above)
Comment (chandanchowdhury) [reply docs/root/modules/snipeit/config.md:9]: Snipe-IT API token uses the permission of the user who generated it: https://snipe-it.readme.io/reference/api-permissions

File: cartography/intel/snipeit/util.py (hunk 2 above)
Comment (chandanchowdhury) [reply cartography/intel/snipeit/util.py:19]: Correct! Credentials are passed via `Authorization` headers (line-25 below)

File: cartography/intel/snipeit/util.py (hunk 3 above)
Comment (chandanchowdhury) [reply cartography/intel/snipeit/util.py:15]: Tried adding the `backoff` decorator as suggested and got below error
```cartography/intel/snipeit/util.py:16: error: Untyped decorator makes function "call_snipeit_api" untyped  [misc]```

File: tests/integration/cartography/intel/snipeit/test_snipeit_assets.py (hunk 7 above)
Comment (chandanchowdhury) [reply tests/integration/cartography/intel/snipeit/test_snipeit_assets.py:56]: I will leave these for now (following same pattern for all tests) and will keep in mind for future tests.

File: docs/root/modules/snipeit/config.md (hunk 4 above)
Comment (achantavy) [reply docs/root/modules/snipeit/config.md:9]: Would be helpful to include that in the doc but I won't block on that

File: docs/root/modules/snipeit/schema.md (hunk 6 above)
Comment (achantavy) [reply docs/root/modules/snipeit/schema.md]: @chandanchowdhury can you confirm that this is correct?

File: docs/root/modules/snipeit/schema.md (hunk 5 above)
Comment (achantavy) [reply docs/root/modules/snipeit/schema.md]: @chandanchowdhury can you confirm that this is correct?

**Review by achantavy — APPROVED** [review]

**Pull request merged**